client_admin/pulp/client/admin/admin_auth.py

```diff
@@ -38,24 +43,47 @@ def __init__(self, context):
 
         self.context = context
 
-        self.create_option('--username', _('server account username'), aliases=['-u'], required=True)
+        self.create_option('--username', _('server account username'), aliases=['-u'], required=False)
         self.create_option('--password', _('server account password'), aliases=['-p'], required=False)
 
     def login(self, **kwargs):
-        # Query the server
-        username = kwargs['username']
-        password = kwargs['password']
-
-        # Hidden, interactive prompt for the password if not specified
-        if password is None:
-            prompt_msg = 'Enter password: '
-            password = self.context.prompt.prompt_password(_(prompt_msg))
-            if password is self.context.prompt.ABORT:
-                self.context.prompt.render_spacer()
-                self.context.prompt.write(_('Login cancelled'))
-                return os.EX_NOUSER
-
-        result = self.context.server.actions.login(username, password).response_body
+        for auth_type in ['Kerberos', 'Basic']:
+            if auth_type is 'Kerberos':
+                if krb is not None and self.has_kerberos_ticket():
+                    __, krb_context = krb.authGSSClientInit("HTTP@%s" % self.context.server.server_info.server.host)
+                    krb.authGSSClientStep(krb_context, "")
+                    negotiate_details = krb.authGSSClientResponse(krb_context)
+                    #logger.info("KERBEROS selected - %s" % krbV.default_context().default_ccache().principal())
+                    username = krbV.default_context().default_ccache().principal()
+                    password = None
+                else:
+                    continue
+            if auth_type is 'Basic':
+                # Query the server
+                username = kwargs['username']
+                password = kwargs['password']
+                # Hidden, interactive prompt for the password if not specified
+
+                if username is None:
+                    self.context.prompt.write(_('No username provided'))
+                    return os.EX_NOUSER
+
+                if password is None:
+                    prompt_msg = 'Enter password: '
+                    password = self.context.prompt.prompt_password(_(prompt_msg))
+                    if password is self.context.prompt.ABORT:
+                        self.context.prompt.render_spacer()
+                        self.context.prompt.write(_('Login cancelled'))
+                        return os.EX_NOUSER
+           
+            #result = self.context.server.actions.login(username, password).response_body
+            try:
+                response = self.context.server.actions.login(username, password)
+                break
+            except Exception:
+                continue
```

Comment:
It seems like we should make sure that at least one of these two were successful. If neither were successful, the loop will still exit and the later code will execute next. This may explain why I saw issues when I had no authentication method.